md-proto/WebAssembly.md

```diff
@@ -0,0 +1,1884 @@
+format, positions are represented with a special syntax.
+
+
+## Validation
+
+### Module Validation
+
+TODO: describe module validation
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries containing
+    - A [control-flow type](#control-flow-type).
+    - A *depth* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During validation, if either stack is empty when an element is to be popped from
+it, validation fails.
+
+#### Control Flow Type
+
+A *control-flow type* is either a [type](#types), `any`, which satisfies any
+type check, or `void`, which satisfies no check.
+
+> `any` is effectively serving as a kind of bottom type, and `void` is
+effectively serving as a kind of top type. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Control Flow Type Merge
+
+To merge two [control-flow types]:
+ - If either control-flow type is `any`, the result is the other control flow
+   type.
+ - Otherwise, if the control-flow types are the same, the result is that control
+   flow type.
+ - Otherwise, validation fails with a type error.
+
+#### Function Validation Initialization
+
+The current position starts at the first position. The type stack begins empty.
+The control-flow stack starts with one entry. This entry's control-flow type is
+the first return type of the function, if it has any, or `void` otherwise. Its
+block depth is zero.
+
+#### Function Body Validation
+
+If the current position is the return position, function return validation is
+prompted and validation of the function is thereafter complete.
+
+Otherwise, the instruction at the current position is remembered, the current
+position is incremented to point to the instruction following it, or if there
+are no instructions following it, to the return position. Then the remembered
+instruction is validated as follows.
+
+If the instruction has a **Validation** clause, that clause describes its
+validation. Otherwise, the following generic validation is performed:
+
+For each operand [type] in the instruction's signature, a type is popped from
+the type stack and checked to be the same as the corresponding operand type.
+Each return type of the instruction's signature is then pushed onto the type
+stack.
+
+If the instruction's signature has no return types, and the size of the value
+stack is greater than the control-flow stack's top's depth, validation fails
+with a type error.
+
+Then, [validation](#function-body-validation) is resumed.
+
+#### Function Return Validation
+
+When a *return* is prompted, one [type] for each type in the function
+signature's return type is popped from the type stack and type-checked against
+its corresponding return type. The type stack is then checked to be empty. If
+there are sufficient stack elements and all checks have passed, function
+validation is successful.
+
```

Comment:
The description does not make it clear if 'type stack is then checked to be empty' is not true then is it a validation failure, and if not then why bother checking?